core/pkg/artifacts/saver.go

```diff
@@ -512,31 +513,115 @@ func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
 		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
 	}
 
-	file, err := os.Open(path)
+	parts, err := computeMultipartHashes(logger, path, getChunkSize(fileSize), runtime.NumCPU())
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		_ = file.Close()
-	}()
+	return parts, nil
+}
 
-	partsInfo := []gql.UploadPartsInput{}
-	partNumber := int64(1)
-	buffer := make([]byte, getChunkSize(fileSize))
-	for {
-		bytesRead, err := file.Read(buffer)
-		if err != nil && err != io.EOF {
-			return nil, err
-		}
-		if bytesRead == 0 {
-			break
+func computeMultipartHashes(logger *observability.CoreLogger, path string, chunkSize int64, numWorkers int) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+
+	if fileSize < chunkSize {
+		return nil, fmt.Errorf("file size is less than chunk size: %d < %d", fileSize, chunkSize)
+	}
+	if numWorkers < 1 {
```

Comment:
this check can come earlier in case it is wrong we don't need to do the stats check,